egs/ami/ASR/pruned_transducer_stateless7/decode.py

```diff
@@ -0,0 +1,681 @@
+        default="greedy_search",
+        help="""Possible values are:
+          - greedy_search
+          - beam_search
+          - modified_beam_search
+          - fast_beam_search
+          - fast_beam_search_nbest
+          - fast_beam_search_nbest_oracle
+          - fast_beam_search_nbest_LG
+        If you use fast_beam_search_nbest_LG, you have to specify
+        `--lang-dir`, which should contain `LG.pt`.
+        """,
+    )
+
+    parser.add_argument(
+        "--beam-size",
+        type=int,
+        default=4,
+        help="""An interger indicating how many candidates we will keep for each
+        frame. Used only when --decoding-method is beam_search or
+        modified_beam_search.""",
+    )
+
+    parser.add_argument(
+        "--beam",
+        type=float,
+        default=4,
+        help="""A floating point value to calculate the cutoff score during beam
+        search (i.e., `cutoff = max-score - beam`), which is the same as the
+        `beam` in Kaldi.
+        Used only when --decoding-method is fast_beam_search""",
+    )
+
+    parser.add_argument(
+        "--ngram-lm-scale",
+        type=float,
+        default=0.01,
+        help="""
+        Used only when --decoding_method is fast_beam_search_nbest_LG.
+        It specifies the scale for n-gram LM scores.
+        """,
+    )
+
+    parser.add_argument(
+        "--max-contexts",
+        type=int,
+        default=8,
+        help="""Used only when --decoding-method is
+        fast_beam_search, fast_beam_search_nbest, fast_beam_search_nbest_LG,
+        and fast_beam_search_nbest_oracle""",
+    )
+
+    parser.add_argument(
+        "--max-states",
+        type=int,
+        default=64,
+        help="""Used only when --decoding-method is
+        fast_beam_search, fast_beam_search_nbest, fast_beam_search_nbest_LG,
+        and fast_beam_search_nbest_oracle""",
+    )
+
+    parser.add_argument(
+        "--context-size",
+        type=int,
+        default=2,
+        help="The context size in the decoder. 1 means bigram; " "2 means tri-gram",
+    )
+    parser.add_argument(
+        "--max-sym-per-frame",
+        type=int,
+        default=1,
+        help="""Maximum number of symbols per frame.
+        Used only when --decoding_method is greedy_search""",
+    )
+
+    parser.add_argument(
+        "--num-paths",
+        type=int,
+        default=200,
+        help="""Number of paths for nbest decoding.
+        Used only when the decoding method is fast_beam_search_nbest,
+        fast_beam_search_nbest_LG, and fast_beam_search_nbest_oracle""",
+    )
+
+    parser.add_argument(
+        "--nbest-scale",
+        type=float,
+        default=0.5,
+        help="""Scale applied to lattice scores when computing nbest paths.
+        Used only when the decoding method is fast_beam_search_nbest,
+        fast_beam_search_nbest_LG, and fast_beam_search_nbest_oracle""",
+    )
+
+    add_model_arguments(parser)
+
+    return parser
```

Comment:
Could you use the decode.py from librispeech?

https://github.com/k2-fsa/icefall/blob/500792d0f1d8da67c1bd70b3066c78cf36811ce3/egs/librispeech/ASR/pruned_transducer_stateless7/decode.py#L173
the above option is crucial for getting better results in decoding.

Reply:
I see. Let me try decoding with this option.

Reply:
Using this gives me slightly worse WERs.

Reply:
What `--avg` value are you using?

Reply:
I used --iter 105000, --avg 9.

Reply:
Have you tried other values of `--avg` instead of `9`. Also, have you tried other combinations of `--epoch` and `--avg`?

Reply:
Ok, you are using `--iter`. Maybe you can also try `--epoch`.

Since you only train 15 epochs, `--avg 9` may be too large.

Reply:
> Using this gives me slightly worse WERs.

This is the first time that I have seen that it causes a worse WER, I think.

Reply:
I will try more combinations and report here today. I tried that one because it was giving me the best WERs without the --use-averaged-model, but perhaps it may not be optimal when used with this option.

Reply:
I tried using different number of epoch checkpoints for averaging, and here are the results using fast beam search. The first row is without `--use-averaged-model`. It seems averaging 7, 8, or 9 of the last epochs gives improvements.

| **Decoding method**                   | **Dev IHM** | **Test IHM** | **Dev SDM** | **Test SDM** | **Dev GSS** | **Test GSS** |
|---------------------------------------|-------------|--------------|-------------|--------------|-------------|--------------|
| FBS (ckpt=105000, avg=10)             |    19.46    |     18.35    |    31.14    |     32.52    |    22.45    |     23.38    |
| FBS (epoch=14, avg=5, use-avg-model)  |    19.75    |     18.63    |    31.27    |     32.26    |     22.7    |     23.58    |
| FBS (epoch=14, avg=6, use-avg-model)  |    19.52    |     18.43    |    31.13    |     32.07    |    22.47    |     23.24    |
| FBS (epoch=14, avg=7, use-avg-model)  |    **19.42**    |     18.17    |   **31.01**    |     **32.04**    |     22.3    |     22.98    |
| FBS (epoch=14, avg=8, use-avg-model)  |    19.44    |     18.04    |    31.11    |     32.1     |    **22.21**    |     22.83    |
| FBS (epoch=14, avg=9, use-avg-model)  |    19.49    |     **17.99**    |    31.34    |     32.35    |    22.22    |     **22.78**    |
| FBS (epoch=14, avg=10, use-avg-model) |    19.55    |     18.04    |    31.85    |     32.88    |    22.41    |     23.03    |

Reply:
I will update the results to use `--epoch 14 --avg 8 --use-averaged-model` since it seems to provide consistent improvements across all sets.

Reply:
Done.